warp10/src/main/java/io/warp10/script/functions/TOGTS.java

```diff
@@ -16,98 +16,257 @@
-    
+
+    // if there is no type map, the output is a map of GTS (v2.3.0 signature), or a map of lists of GTS
+    if (null == typeMap) {
+      Map<String, ArrayList<GeoTimeSerie>> result = new HashMap<String, ArrayList<GeoTimeSerie>>();
+      for (GTSDecoder decoder: decodersInput) {
+        Map<String, GeoTimeSerie> series = new HashMap<String, GeoTimeSerie>();
+        GeoTimeSerie gts;
+        while (decoder.next()) {
+          Object value = decoder.getBinaryValue();
+
+          String type = "DOUBLE";
+
+          if (value instanceof String) {
+            type = "STRING";
+          } else if (value instanceof Boolean) {
+            type = "BOOLEAN";
+          } else if (value instanceof Long) {
+            type = "LONG";
+          } else if (value instanceof Double || value instanceof BigDecimal) {
+            type = "DOUBLE";
+          } else if (value instanceof byte[]) {
+            type = "BINARY";
+          }
+
+          gts = series.get(type);
+
+          if (null == gts) {
+            gts = new GeoTimeSerie();
+            gts.setMetadata(decoder.getMetadata());
+            series.put(type, gts);
+          }
+
+          GTSHelper.setValue(gts, decoder.getTimestamp(), decoder.getLocation(), decoder.getElevation(), value, false);
+        }
+        // exit here if input is not a list.
+        if (!listInput) {
+          stack.push(series);
+          return stack;
+        }
+        // merge the series Map into the big one
+        for (Entry<String, GeoTimeSerie> entry: series.entrySet()) {
+          if (!result.containsKey(entry.getKey())) {
+            result.put(entry.getKey(), new ArrayList<GeoTimeSerie>());
+          }
+          result.get(entry.getKey()).add(entry.getValue());
+        }
+      }
+      stack.push(result);
+      return stack;
+
+    } else {
+      // if there is a type map on the stack:
+      //  - the map key is a selector. If the encoder's metadata fits, the gts will have the corresponding type of the first matching selector.
+      //  - if the encoder doesn't fit to any selector, the gts will have the type of the first encountered element in the encoder.
+      // GTSHelper.setValue will try to convert values whenever possible, and
+      // a byte array will be serialized as an ISO-8859-1 string by GTSHelper.setValue
+      ArrayList<GeoTimeSerie> result = new ArrayList<GeoTimeSerie>();
+
+      for (GTSDecoder decoder: decodersInput) {
+        GeoTimeSerie gts = new GeoTimeSerie();
+        gts.setMetadata(decoder.getMetadata());
+        String enforcedType = null;
+        for (Entry<String, ArrayList<MetadataSelectorMatcher>> entry: typeMap.entrySet()) {
+          for (MetadataSelectorMatcher m: entry.getValue()) {
+            if (m.MetaDataMatch(decoder.getMetadata())) {
+              enforcedType = entry.getKey();
+            }
+            if (null != enforcedType) {
+              break;
```

Comment:
This only break the for on the list of `MetadataSelectorMatcher`, not for all `MetadataSelectorMatcher` for all types.
If several MetadataSelectorMatcher match, the behavior is quite strange.
I think you should check if `enforcedType` is already set and throw an error if it is the case, to avoid GTS matching several types.
Here is a minimal example for this issue:
```
NEWENCODER
{ 
  'BOOLEAN' '~.*{}'
  'STRING' '~.*{}' 
} 
->GTS
```

Reply:
Indeed, I must break the type loop also. 
I do not want to raise an error. Order matters in the input map, that's why I used a linkedHashMap.
If user want a fallback to another type, it could be done with 
```
{ 'LONG' 'foo{unit=42}' 
  'DOUBLE' '~.*{}'
} ->GTS
```